enhancements/substitutes-for.md

```diff
@@ -0,0 +1,379 @@
+into "skips", but there is an argument for instead moving the existing
+"replaces" reference to be a skips, and making the "replaces" reference
+point to the "newest" substitution.  That would at least ensure that
+if/when we must install a bundle as part of the path to upgrade to the
+latest, we install the most recently patched version of that bundle.
+
+However it adds additional complexity to determine which bundle is the
+"latest rebuild" of a given bundle, and more mutation to the CSV.
+
+I'd like to further discuss the implications of not doing this and what
+scenarios might be realistically problematic.
+
+### Risks and Mitigations
+
+The primary risk is that in mutating the CSVs we break the upgrade path in some
+way, either by not adding skips/replaces that we should, adding incorrect ones,
+or changing the upgrade behavior (such as the aforementioned cases where
+we go from a "replaces" upgrade path to a "skips" upgrade path, and some operator
+migration logic does not get executed).
+
+Not aware of any security concerns.
+
+
+## Design Details
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+Need to test the user scenarios described above.  Generally:
+
+1) rebuilding a bundle multiple times (creating transitive substitution
+relationships)
+2) rebuilding bundles that are referenced by other bundles
+
+And ensuring in both cases that the CSV content served by ListBundles
+contains all the necessary/correct skips/replaces references to ensure
+a correct upgrade graph for all channels.
+
+### Graduation Criteria
+
+This will be put directly into production once implemented, as it will be
+consumed by freshmaker.  So substitutes-for will need to be a GA api with
+a GA implementation in OPM.
+
+### Upgrade / Downgrade Strategy
+
+All logic related to this behavior will be tied to the opm binary used to
+generate the catalog, and the opm binary in the catalog image that is used
+to serve the content.  Therefore it should not require any changes to 
+OLM on the target cluster.  Catalogs produced that use this feature must
+work with any existing supported OCP cluster.
+
+Users who want to take advantage of this feature will need to use a current
+opm binary both to build their catalog DB as well as packaged in their 
+catalog image.
+
+Note:  Today the operator pipeline uses the latest opm binary to build catalog DBs
+(good) but matches the opm binary in the catalog to the ocp version.  We will
+need to either:
+
+1) Convince ourselves it is safe to always use a current opm binary in the catalog,
+even when the catalog is targeted at older clusters
+or
+2) Backport the substitutes-for support in opm to all opm versions needed for
+the operator pipeline (today that means back to 4.5 at least).
+
+If an existing catalog DB is being modified by this new opm binary, it will need
+to add the new table/column tracking the substiutes-for metadata.
```

Comment:
This would just be a db migration, presumably. That's how historical changes to the database have always worked.

Reply:
ack, i thought you guys already had a process for this, just wanted to explicitly call out as many of the implementation steps as i could so nothing got missed.